### PR TITLE
hackage-db: support cabal-install 3.10.1.0 XDG paths.

### PR DESCRIPTION
**Pull request by athas**
From 3.10.1.0 onwards, cabal-install uses XDG directories to store its state and cache components.  cabal2nix expects to find the Hackage tarball in ~/.cabal, where it may no longer be located.

This commit changes cabal2nix to look in ~/.cabal if that directory exists (which matches the behaviour of cabal-install itself), but otherwise looks in the appropriate XDG directory.

### CODE REVIEW

**Comment by sternenseemann** [comment]
Thanks! Can you also link to the [upstream code](https://github.com/haskell/cabal/blob/0ed12188525335ac9759dc957d49979ab09382a1/cabal-install/src/Distribution/Client/Config.hs#L594-L610) that implements this in the comment for reference?

This also shows that we should respect the `CABAL_DIR` environment variable, but we can also do that in a separate change, as we weren't even respecting it in the first place.

**Comment by athas** [comment]
Link added.

The Right Solution is perhaps to depend on `cabal` to use exactly the same logic.  However, cabal2nix got away with not doing exactly what `cabal` does so far (in particular, not respecting `CABAL_CONFIG` or `CABAL_DIR`), so it probably does not matter in practice.

**Comment by sternenseemann** [comment]
Seems like cabal-install has made our lives harder than it should be – namely you can also configure arbitrary location for the package db location. What's worse this is apparently rendered into the configuration file by default, so even if people move their config to xdg paths, it may still show up under `~/.cabal`.

Not *quite* sure yet how to proceed – we'll want to avoid depending on `cabal-install` I think, since it is quite the dependency. One nice thing is that we support a wide range of `Cabal` libraries, but supporting a wide range of `cabal-install` is likely very painful.

**Comment by athas** [comment]
The issue with paths being stored in the cabal config file is not new.  It's been like that for a while - maybe forever.  This PR does not make the situation worse.

**Comment by sternenseemann** [comment]
True, but it'll be more likely to be a problem now. In any case that we'll have to prefer `~/.cabal` anyways will get us out of 90% of the issues hopefully.

For anything else I'll implement `CABAL_DIR` and people will have to work around it manually.

File: hackage-db/src/Distribution/Hackage/DB/Path.hs [hunk 1]
```diff
@@ -39,9 +50,11 @@ hackageTarballDir :: IO FilePath
 hackageTarballDir = cabalTarballDir "hackage.haskell.org"
 
 -- | Determine the default path of the Hackage database, which typically
--- resides at @"$HOME\/.cabal\/packages\/hackage.haskell.org\/00-index.tar"@.
--- Running the command @"cabal update"@ will keep that file up-to-date.
-
+-- resides in @$HOME\/.cabal\/packages\/hackage.haskell.org\/@.
```
Comment (maralorn) [inline hackage-db/src/Distribution/Hackage/DB/Path.hs:59]: Do we want to adapt this path to the new default?

File: hackage-db/src/Distribution/Hackage/DB/Path.hs [hunk 2]
```diff
@@ -15,10 +15,37 @@ import Distribution.Hackage.DB.Errors
...
+cabalStateDir = do
+  envCabal <- lookupEnv "CABAL_DIR"
+  dotCabal <- getAppUserDataDirectory "cabal"
+  dotCabalExists <- doesDirectoryExist dotCabal
+  case envCabal of
+    Just dir -> pure dir
+    Nothing ->
+      if dotCabalExists
+      then pure dotCabal
+      else getXdgDirectory XdgCache "cabal"
```
Comment (maralorn) [inline hackage-db/src/Distribution/Hackage/DB/Path.hs:48]:
```suggestion
    Nothing | dotCabalExists -> pure dotCabal
    Nothing -> getXdgDirectory XdgCache "cabal"
```
Really unimportant. Just suggesting it to learn whether you also like it better and if not, why not.

**Review by maralorn — APPROVED** [review]
Excellent! I left to small comments, but can also be merged as is.

File: hackage-db/src/Distribution/Hackage/DB/Path.hs (hunk 2 above)
Comment (athas) [reply hackage-db/src/Distribution/Hackage/DB/Path.hs:48]: This was copy-pasted from the cabal source code which doesn't seem to use guards much.

**Pull request merged**